### ISSUE TITLE
[go_router] Add pathParameters and queryParameters to go() 

### ISSUE THREAD

**Issue by johnpryan** [flutter/flutter#108173]
Right now, users rely go `goNamed` to fill in path and query parameters:

```dart
context.go(context.namedLocation('user', params: {'userId': 123}))
```

But there is no way to fill in path or query parameters using go():

```dart
final uri = Uri.parse('/user/${42}').replace(
    queryParameters: {
      'filter': 'abc'
    });
context.go(uri.toString());
```

Instead, go() could provide two optional parameters, and `pathParameters` and `queryParameters` and fill in the template on behalf of the user:

```dart
context.go('/user/:id',
    pathParameters: {'id': '123'},
    queryParameters: {'filter': 'abc'});
```

Context: https://github.com/flutter/flutter/issues/107729#issuecomment-1185982122

**Comment by chunhtai** [comment]
I am not entirely sure this is the direction we want to go given, because I think the named location is a cleaner API IMO. Is there a use case where named location may not be preferable?

**Comment by johnpryan** [comment]
This is for users who don't want to used the `namedLocation` API

**Comment by chunhtai** [comment]
but why would one prefer this over namedLocation

**Comment by johnpryan** [comment]
Discussed offline, we don't want to be re-implementing things that Dart core libraries (like Uri) do already. We could consider making go() take a Uri, but we'll just be converting it back into a string right away. So we are going to recommend users use Uri to construct their path if they aren't using the named location API.  We also decided [not to remove the namedLocation API](https://github.com/flutter/flutter/issues/107729).

We also discussed the possibility of `go()` accepting a GoRoute with path and query parameters, so that users don't need to use raw strings, but this isn't something we're going to work on right now. 

I find the namedLocation API strange. It exists in other Flutter routing packages such as vrouter, but isn't a common way of navigating in Angular or React Router.
